### DISCUSSION TITLE
Slow insert speed

### DISCUSSION THREAD

**Discussion by dee-corr** [milvus-io/milvus#18973] [Q&A and General discussion] [answered]
Hi everyone,

I'm new to Milvus and am experiencing slow insert speed. I'm using Milvus standalone on a r6a.8xlarge ec2 instance and trying to upload 24,000,000 records containing a string field and a binary vector embedding. It is a bit vector with a dim in 512, but milvus only accepts it if I set the dim to 4096.

Here is the schema definition:

fields = [
        FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
        FieldSchema(name="smiles", dtype=DataType.VARCHAR, max_length=200),
        FieldSchema(
            name="morgan_fingerprint",
            dtype=DataType.BINARY_VECTOR,
            dim=4096,
            is_primary=False,
        ),
    ]
schema = CollectionSchema(fields, collection_name)
collection = Collection(collection_name, schema, consistency_level="Strong")

I'm experiencing slow insert speed which I'm guessing is to do with the high dimensionality. I'm inserting approx 50,000 records each time which seems to take about ten minutes to complete.

Here is a snapshot of the docker stats during upload:
<img width="886" alt="image" src="https://user-images.githubusercontent.com/51026617/187906216-68da81ca-a25d-4f24-83c5-0903ee39893f.png">

Could you tell me what is wrong here? Is it just too much data in one request or is there another reason for the latency? From the docer stats the network IO between milvus and minio looks very high.

Thanks!

**Comment by yhmo** [comment, marked as answer]
I test two scripts on my local, one script inserts float vectors, the other inserts binary vectors. The performance of inserting binary vectors is indeed poor.

I think this is a performance issue, I have created an issue for investigation: https://github.com/milvus-io/milvus/issues/18985

**Reply by dee-corr** [reply]
Thank you @yhmo for your help!

**Comment by yhmo** [comment]
@dee-corr 
I have made a fix on python SDK to improve the insert performance for binary vectors: https://github.com/milvus-io/pymilvus/pull/1134/files
After this fix, the time cost of inserting 50000 vectors with 4096 dim will reduce from 7 ~ 10 minutes to about 100 milliseconds.
This fix will be available on the next minor version of pymilvus. 
It is only one line change. You also can find out your local pymilvus lib path and modify this file: **pymilvus/client/entity_helper.py**. Replace the two lines:
```
for vector in entity.get("values"):
    field_data.vectors.binary_vector += vector
```
to be:
```
field_data.vectors.binary_vector = b''.join(entity.get("values"))
```

**Comment by dee-corr** [comment]
Hi @yhmo . I will do that. We're very impressed with how responsive and helpful you guys have been.

Thank you!
